daemons/ipa-sam/ipa_sam.c

```diff
@@ -3554,6 +3554,77 @@ static NTSTATUS getsam_interdom_trust_account(struct pdb_methods *methods,
 	return status;
 }
 
+static NTSTATUS ipasam_getsampwsid(struct pdb_methods *methods,
+				    struct samu *user,
+				    const struct dom_sid *sid)
+{
+	struct ipasam_private *ipasam_state =
+			talloc_get_type_abort(methods->private_data, struct ipasam_private);
+	TALLOC_CTX *tmp_ctx;
+	NTSTATUS status;
+	char *filter = NULL;
+	char *sid_str = NULL;
+	LDAPMessage *result = NULL;
+	LDAPMessage *entry = NULL;
+	int ret;
+	int count;
+
+	tmp_ctx = talloc_new(NULL);
+	if (tmp_ctx == NULL) {
+		return NT_STATUS_NO_MEMORY;
+	}
+
+	sid_str = sid_talloc_string(ipasam_state->idmap_ctx, tmp_ctx, sid);
+	if (sid_str == NULL) {
+		return NT_STATUS_NO_MEMORY;
+	}
+
+	filter = talloc_asprintf(tmp_ctx, "(&(|(%s=%s)(%s=%s))(%s=%s))",
+					  LDAP_ATTRIBUTE_OBJECTCLASS,
+					  LDAP_OBJ_SAMBASAMACCOUNT,
+					  LDAP_ATTRIBUTE_OBJECTCLASS,
+					  LDAP_OBJ_ID_OBJECT,
+					  LDAP_ATTRIBUTE_SID, sid_str);
+	if (filter == NULL) {
+		status = NT_STATUS_NO_MEMORY;
+		goto done;
+	}
+
+	ret = smbldap_search(ipasam_state->ldap_state,
+			     ipasam_state->base_dn,
+			     LDAP_SCOPE_SUBTREE,filter, NULL, 0,
+			     &result);
+	if (ret != LDAP_SUCCESS) {
+		status = NT_STATUS_NO_SUCH_USER;
+		goto done;
```

Comment:
Is this safe? What about search timeouts, ACI and other sorts of problems?

Reply:
We have the same code in `ipasam_getsampwnam()` and other places for years.

Reply:
What caught my eye here was another ticket which I can't find now but it was related to IPA incorrectly returning NO_SUCH_USER (or notfound) in a 389-ds plugin when it was I think related to a timeout. IIRC there was a bit of back and forth before Sumit finally discovered the issue.

Reply:
Ah, no that issue was totally different. That was extop plugin action which is triggered by the use of extended operation by SSSD. It is not the same as we have here. I don't think it is worth the check here. If `cifs/..` principal loses ACIs, we'll notice it earlier because we rely on a number of other ACIs here assigned to the trust agent group.